### PR TITLE
Testing in docker container

### PR DESCRIPTION
**Pull request by supermomonga**
# Description

I wrote `Dockerfile` and `docker-compose.yml` to run tests in docker container.

I also wrote an instruction about testing into README. I think my english is poor, so I wish someone to rewrite it later.

# Purpose

Main purpose is to get completely isolated environment to run tests.

I motivated to do this because I couldn't run `run-integration-test.sh` in my env.

### CODE REVIEW

**Comment by razzmatazz** [comment]
This looks good! And it works for me too.

Just two notes:
 - could you modify `travis.yaml` so it would run testing via this docker procedure too?
 - and then probably update existing test*.sh files to invoke docker-compose build* instead?

..as it is better to have a unified testing procedure rather than 2 different ones..

And your English is most probably better than mine, so we'll wait for some native speakers to upload a PR :)

**Comment by supermomonga** [comment]
> could you modify travis.yaml so it would run testing via this docker procedure too?

Yeah, I'll dig travis document to find out how to cache docker image between builds.

---

>and then probably update existing test*.sh files to invoke docker-compose build* instead?

I think there is two option.

1. Recommend to run `docker-compose` command manually.
    - `docker-compose up` can run all tests. `docker-compose run --rm [test-kind]` can run one kind of tests.
    - Any Docker supported OS can run tests (include Windows).
2. Create `run-all-tests.sh` for *nix based OS, and `run-all-tests.ps1` for Windows OS.
    - Those files are just wrapper of `docker-compose` command.

I think options 1 is better, because option 2 will increase complexity by adding unnecessarily layer (bash/powershell script).

And I'm worry about forcing to use (install) Docker to all developers.
Maybe we should keep current testing way (run tests in host machine) and provide Docker container as optional way?

---

>..as it is better to have a unified testing procedure rather than 2 different ones..

It's nice that unifying unit tests and integration tests into one procedure.
Should I also unify melpa building test into that?

**Comment by supermomonga** [comment]
I added `[WIP]` to the title 🕺

**Comment by razzmatazz** [comment]
- I would actually prefer option 2 (wrapper scripts) -- there are not that many devs, and unification is better imho; easier to type/autocomplete and you don't have to read README to run them...
 - dependency on docker for development is not unreasonable, imho;
 - actually we already had an issue just for this, https://github.com/OmniSharp/omnisharp-emacs/issues/290
 - maybe melpa testing can go in too? not sure, if it is easy for you to do?

**Comment by supermomonga** [comment]
> - I would actually prefer option 2 (wrapper scripts) -- there are not that many devs, and unification is better imho; easier to type/autocomplete and you don't have to read README to run them...
> - dependency on docker for development is not unreasonable, imho;
> - actually we already had an issue just for this, #290

I think you are right. That makes testing procedure more easier and clearer, and I totally agree that it's important.

OK, I'll create a wrapper script.

> maybe melpa testing can go in too? not sure, if it is easy for you to do?

Yes, I think I can. I'll also do this.

**Comment by supermomonga** [comment]
@razzmatazz I think all tasks completed now 😃 

---

# Changes

1. All tests will be executed in docker container at Travis.
2. Better melpa building test
    - Works with forked repository's travis integration.
      - Before: It tries to get branch from omnisharp org's official repository, so it didn't work
      - After: It tries to get branch from forked repository, so It works file.
    - Works with local environment.
      - Before: It always uses remote repository's code, so local changes will not be applied.
      - After: It uses local git repository's code, so local changes will be applied.
3. We simply need to run `run-tests.sh` to execute all kind of tests.

I couldn't create `run-tests.ps1` because I don't have Windows machine, so can I leave it for now?
We can wait a Pull Request from someone who uses Windows machine.

File: run-tests.sh [hunk 1]
```diff
@@ -1,5 +1,20 @@
-#! /bin/bash
-# Runs all tests for the project
+#! /bin/bash -e
+set -o pipefail
+
+check_command_exists () {
```
Comment (sp3ctum) [inline run-tests.sh]: Nice!

**Comment by supermomonga** [comment]
Ah, I forgot to update README.md. Done 💦

**Comment by razzmatazz** [comment]
This looks and works great for me.
Thanks a lot!

**Pull request merged**